plugins/ArchivingMetrics/Timer.php

```diff
@@ -0,0 +1,186 @@
+
+namespace Piwik\Plugins\ArchivingMetrics;
+
+use Piwik\ArchiveProcessor\Rules;
+use Piwik\Plugins\ArchivingMetrics\Clock\Clock;
+use Piwik\Plugins\ArchivingMetrics\Clock\ClockInterface;
+use Piwik\Plugins\ArchivingMetrics\Writer\DbWriter;
+use Piwik\Plugins\ArchivingMetrics\Writer\WriterInterface;
+
+final class Timer
+{
+    /**
+     * @var bool
+     */
+    private $isArchivePhpTriggered;
+
+    /**
+     * @var ClockInterface
+     */
+    private $clock;
+
+    /**
+     * @var WriterInterface
+     */
+    private $writer;
+
+    private $runs = [];
+
+    /**
+     * @var Timer
+     */
+    private static $instance;
+
+    public function __construct(bool $isArchivePhpTriggered, ClockInterface $clock, WriterInterface $writer)
+    {
+        $this->isArchivePhpTriggered = $isArchivePhpTriggered;
+        $this->clock = $clock;
+        $this->writer = $writer;
+    }
+
+    public static function getInstance(bool $isArchivePhpTriggered, ?ClockInterface $clock = null, ?WriterInterface $writer = null): self
+    {
+        if (self::$instance !== null) {
+            return self::$instance;
+        }
+
+        if ($clock === null) {
+            $clock = new Clock();
+        }
+
+        if ($writer === null) {
+            $writer = new DbWriter();
+        }
+
+        self::$instance = new self($isArchivePhpTriggered, $clock, $writer);
+        return self::$instance;
+    }
+
+    public function start(Context $context): void
+    {
+        if (false === $this->isApplicableForTiming($context)) {
+            return;
+        }
+
+        $this->runs[$context->getKey()] = [
+            'idsite' => $context->idSite,
+            'period' => $context->period->getLabel(),
+            'segment' => $context->segment->getString(),
+            'date1' => $context->period->getDateTimeStart()->toString('Y-m-d'),
+            'date2' => $context->period->getDateTimeEnd()->toString('Y-m-d'),
+            'ts_started' => $this->clock->now(),
+            'timeStarted' => $this->clock->microtime(),
+        ];
+    }
+
+    public function complete(Context $context, array $idArchives, bool $wasCached): void
+    {
+        if (false === $this->isApplicableForTiming($context)) {
+            return;
+        }
+
+        if (true === $wasCached || empty($idArchives)) {
+            return;
+        }
+
+        $key = $context->getKey();
+
+        if (!isset($this->runs[$key]['timeStarted'])) {
+            return;
+        }
+
+        $finishedAt = $this->clock->microtime();
+        $totalTimeMs = ($finishedAt - $this->runs[$key]['timeStarted']);
+
+        $this->runs[$key]['totalTime'] = $totalTimeMs;
+        $this->runs[$key]['timeFinished'] = $finishedAt;
+        $this->runs[$key]['ts_finished'] = $this->clock->now();
+
+        $exclusiveTimeMs = $this->calculateExclusiveTime($key);
+        $this->runs[$key]['exclusiveTime'] = $exclusiveTimeMs;
+
+        $this->writer->write([
+            'idarchive' => reset($idArchives),
```

Comment:
Is there any way this will ever be called for more than one `idarchive`? And if so, wouldn't it loose information if it only stores a metric for the first one?